libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,453 @@
+   default:
+      // Since the bar-division hypotheses are based on these very time
+      // signatures, this should never happen. But the code would need to be
+      // structured more clearly to really have confidence and risk crashing.
+      // For now we accept that the time-signature cannot be recognized in
+      // release builds.
+      assert(false);
+      return std::nullopt;
+   }
+}
+
+double
+GetTimeSignatureLikelihood(const std::optional<TimeSignature>& ts, double bpm)
+{
+   // TODO these were taken from a rather old PoC - review.
+   if (!ts.has_value())
+      return 0.;
+
+   static const std::unordered_map<TimeSignature, double> expectedBpms {
+      { TimeSignature::TwoTwo, 115. },
+      { TimeSignature::FourFour, 115. },
+      { TimeSignature::ThreeFour, 140. },
+      { TimeSignature::SixEight, 64. },
+   };
+
+   static const std::unordered_map<TimeSignature, double> bpmStdDevs {
+      { TimeSignature::TwoTwo, 25. },
+      { TimeSignature::FourFour, 25. },
+      { TimeSignature::ThreeFour, 25. },
+      { TimeSignature::SixEight, 15. },
+   };
+
+   // Add a slight bias towards 4/4, which is the most common time signature.
+   static const std::unordered_map<TimeSignature, double> tsPrior {
+      { TimeSignature::TwoTwo, .1 },
+      { TimeSignature::FourFour, .45 },
+      { TimeSignature::ThreeFour, .2 },
+      { TimeSignature::SixEight, .25 },
+   };
+
+   const auto mu = expectedBpms.at(*ts);
+   const auto sigma = bpmStdDevs.at(*ts);
+   const auto tmp = (bpm - mu) / sigma;
+   const auto likelihood = std::exp(-.5 * tmp * tmp);
+   return likelihood * tsPrior.at(*ts);
+}
+
+// To evaluate how likely a certain BPM is, we evaluate how much the ODF repeats
+// itself at that beat rate. We do this by looking at the auto-correlation of
+// the ODF, "comb-filtering" it at integer multiples of the beat period.
+double GetBeatSelfSimilarityScore(
+   double odfAutoCorrSampleRate, double bpm,
+   const std::vector<float>& odfAutoCorr, int odfAutocorrFullSize,
+   QuantizationFitDebugOutput* debugOutput)
+{
+   // Look for closest peak in `odfAutoCorr`:
+   const auto lag = odfAutoCorrSampleRate * 60 / bpm;
+   // Traverse all the auto-correlation by steps of `k*lag`, each time
+   // finding the closest peak, and average the values. Doing this rather
+   // than evaluating only just one peak is more robust to rhythms which
+   // miss beats.
+   auto periodIndex = 1;
+   auto sum = 0.;
+   auto numIndices = 0;
+   while (true)
+   {
+      auto j = static_cast<int>(periodIndex++ * lag + .5);
+      if (j >= odfAutoCorr.size())
+         break;
+      while (true)
+      {
+         const auto i = MapToPositiveHalfIndex(j - 1, odfAutocorrFullSize);
+         const auto k = MapToPositiveHalfIndex(j + 1, odfAutocorrFullSize);
+         if (
+            odfAutoCorr[i] <= odfAutoCorr[j] &&
+            odfAutoCorr[j] >= odfAutoCorr[k])
+            break;
+         j = odfAutoCorr[i] > odfAutoCorr[k] ? i : k;
+      }
```

Comment:
So, this while loop is climbing up to a local peak?  Summarize with a comment if I understand right